### DISCUSSION TITLE
Developer mode: Enables custom frame generation from Z-Wave XML

### DISCUSSION THREAD

**Discussion by silabs-MarkM** [zwave-js/zwave-js-ui#3887] [Ideas]
# Description
Z-Wave XML contains information about all the available command classes.
E.g: List of command classes, command, available fields.
This XML file can be parsed and according to this, on the UI the developer could set the fields.
This behavior would be the same, how the PC Controller acts. On the UI the developer can set the fields of the supported command class and send the raw data to the backend. According to my information (@AlCalzone), the custom frame can be called as:
```
const cc = new CommandClass(driver, {
    nodeId: {nodeID},
    ccId: {CommandClasses},
    ccCommand: {Command},
    payload: Buffer.from([0xXX, 0xXX, 0xXX, ...])
});
await driver.sendCommand(cc);
```
For now, only the "Driver Function"-s provides this feature, when somebody wants to send a custom frame to their device. 

# Example1:
```
<cmd_class key="0x25" version="1" name="COMMAND_CLASS_SWITCH_BINARY" help="Command Class Binary Switch">
    <cmd key="0x02" name="SWITCH_BINARY_GET" help="Switch Binary Get" support_mode="RX" />
    <cmd key="0x03" name="SWITCH_BINARY_REPORT" help="Switch Binary Report" support_mode="TX">
      <param key="0x00" name="Value" type="BYTE">
        <bitflag key="0x01" flagname="off/disable" flagmask="0x00" />
        <bitflag key="0x02" flagname="on/enable" flagmask="0xFF" />
      </param>
    </cmd>
    <cmd key="0x01" name="SWITCH_BINARY_SET" help="Switch Binary Set" support_mode="RX">
      <param key="0x00" name="Switch Value" type="BYTE">
        <bitflag key="0x01" flagname="off/disable" flagmask="0x00" />
        <bitflag key="0x02" flagname="on/enable" flagmask="0xFF" />
      </param>
    </cmd>
 </cmd_class>
```
According to this part of the XML, on the UI the developer can choose the COMMAND_CLASS_SWITCH_BINARY (available version which can be 1 or 1+) and the SWITCH_BINARY_SET command. The UI could generate them one (in this case) field, which could be a dropdown menu with the `on/enable=0xFF` and the `off/disable=0x00` values.

# Example2
Some command classes have a payload of variable length, like the Color Switch CC.

```
<cmd_class key="0x33" version="1" name="COMMAND_CLASS_SWITCH_COLOR" help="Command Class Color Switch">
    <cmd key="0x01" name="SWITCH_COLOR_SUPPORTED_GET" help="Color Switch Supported Get" support_mode="RX" />
    <cmd key="0x02" name="SWITCH_COLOR_SUPPORTED_REPORT" help="Color Switch Supported Report" support_mode="TX">
      <param key="0x00" name="Color Component mask" type="WORD" />
    </cmd>
    <cmd key="0x03" name="SWITCH_COLOR_GET" help="Color Switch Get" support_mode="RX">
      <param key="0x00" name="Color Component ID" type="BYTE" />
    </cmd>
    <cmd key="0x04" name="SWITCH_COLOR_REPORT" help="Color Switch Report" support_mode="TX">
      <param key="0x00" name="Color Component ID" type="BYTE" />
      <param key="0x01" name="Value" type="BYTE" />
    </cmd>
    <cmd key="0x05" name="SWITCH_COLOR_SET" help="Color Switch Set" support_mode="RX">
      <param key="0x00" name="Properties1" type="STRUCT_BYTE">
        <bitfield key="0x00" fieldname="Color Component Count" fieldmask="0x1F" />
        <bitfield key="0x01" fieldname="Reserved" fieldmask="0xE0" shifter="5" />
      </param>
      <variant_group key="0x01" name="vg1" paramOffs="0x00" sizemask="0x1F" sizeoffs="0x00">
        <param key="0x00" name="Color Component ID" type="BYTE" />
        <param key="0x01" name="Value" type="BYTE" />
      </variant_group>
    </cmd>
....
```
When the developer chooses this CC with the SWITCH_COLOR_SET command, the UI could generate the following fields:
- Color Component Count
- Reserved (bit)
And a plus and minus button to add vg-s for the command class. One VG would have the following fields:
- Color Component ID: a form which can store values between 0 and 255
- Value: another form with the same settings.

**Comment by AlCalzone** [comment]
I suppose the Z-Wave XML repo needs to be made public before we could do this.

**Reply by silabs-MarkM** [reply]
Most of the developers have access to the XML repository. If you "browse" file, you can change it dynamically. In this case, I don't see any blocker, what do you think?

**Reply by AlCalzone** [reply]
That could be an option, yes.

**Comment by robertsLando** [comment]
I have created https://github.com/zwave-js/zwave-js-ui/issues/3888 as I think this could live inside a custom plugin
